### PR TITLE
WIP: Add support for NumericValue

### LINKED ISSUE #68: Support `NumericValue` property
This maps, for example, `𝟜` and `௪` to the number 4, or `⅐` to 0.14285714. It would need to return a `Double`, because not all characters have integral values. There could perhaps be a second API optimised for integral values.

The data are in https://www.unicode.org/Public/14.0.0/ucd/extracted/DerivedNumericValues.txt, which has been extracted from https://www.unicode.org/Public/14.0.0/ucd/UnicodeData.txt

### PR DESCRIPTION
**Pull request by wismill**
Fixes #68

### CODE REVIEW

File: lib/Unicode/Char/Numeric.hs [hunk 1]
```diff
@@ -41,3 +45,15 @@ isNumber c = case generalCategory c of
     LetterNumber  -> True
     OtherNumber   -> True
     _             -> False
+
+
```
Comment (harendra-kumar) [inline lib/Unicode/Char/Numeric.hs]: Remove additional blank line.

File: lib/Unicode/Char/Numeric.hs [hunk 2]
```diff
@@ -41,3 +45,15 @@ isNumber c = case generalCategory c of
...
+-- | Numeric value of a character, if relevant.
+{-# INLINE numericValue #-}
+numericValue :: Char -> Maybe Rational
+numericValue = V.numericValue
+
+-- | Integer value of a character, if relevant.
+intValue :: Char -> Maybe Int
+intValue c = V.numericValue c >>= \r -> if denominator r == 1
+    then Just (fromIntegral (numerator r))
+    else Nothing
```
Comment (harendra-kumar) [inline lib/Unicode/Char/Numeric.hs:64]: Generally the last line in a file ends with a newline, github warns about it.

File: lib/Unicode/Char/Numeric.hs [hunk 3]
```diff
@@ -41,3 +45,15 @@ isNumber c = case generalCategory c of
...
+
+
+-- | Numeric value of a character, if relevant.
+{-# INLINE numericValue #-}
+numericValue :: Char -> Maybe Rational
+numericValue = V.numericValue
+
+-- | Integer value of a character, if relevant.
+intValue :: Char -> Maybe Int
+intValue c = V.numericValue c >>= \r -> if denominator r == 1
```
Comment (harendra-kumar) [inline lib/Unicode/Char/Numeric.hs:62]: Would be better if the `if` statement is on the next line. Even better if a `do` block is used.

Comment (harendra-kumar) [reply lib/Unicode/Char/Numeric.hs:62]: We can use the longer name `integerValue` instead, it is no longer than `numericValue` anyway.

**Review by harendra-kumar — COMMENTED** [review]
Looks good. Just some nitpicks.

File: test/Unicode/CharSpec.hs [hunk 4]
```diff
@@ -77,6 +80,14 @@ spec = do
       UChar.toUpper `shouldBeEqualTo` Char.toUpper
     it "toTitle" do
       UChar.toTitle `shouldBeEqualTo` Char.toTitle
+  describe "Numeric" do
+    it' "isNumber" do
+      UNumeric.isNumber `shouldBeEqualTo` Char.isNumber
+    it "isNumber implies a numeric value" do
+      -- [NOTE] the following does not hold with the current predicate `isNumber`.
+      -- 'let check c = (UNumeric.isNumber c `xor` isNothing (UNumeric.numericValue c))
+      let check c = not (UNumeric.isNumber c) || isJust (UNumeric.numericValue c)
```
Comment (harendra-kumar) [inline test/Unicode/CharSpec.hs:89]: I wonder how many such cases are there where `isNumber` returns False but the value is numeric. Is it a lot of cases or just a few?

File: lib/Unicode/Char/Numeric.hs [hunk 5]
```diff
@@ -46,13 +46,18 @@ isNumber c = case generalCategory c of
     OtherNumber   -> True
     _             -> False
 
-
 -- | Numeric value of a character, if relevant.
+--
+-- __Note:__ a character may have a numeric value but return 'False' with
+-- the predicate 'isNumber', because 'isNumber' only tests 'GeneralCategory'.
```
Comment (harendra-kumar) [inline lib/Unicode/Char/Numeric.hs:52]: We may want to put a note in `isNumber` as well.

File: test/Unicode/CharSpec.hs (hunk 4 above)
Comment (wismill) [reply test/Unicode/CharSpec.hs:89]: 81 CJK characters:
```haskell
ns = filter (\c -> not (isNumber c `xor`  isNothing (numericValue c))) [minBound .. maxBound]
-- 81
(\c -> (showHex (ord c) "", generalCategory c)) <$> ns
-- [("3405",OtherLetter),("3483",OtherLetter),("382a",OtherLetter),("3b4d",OtherLetter),("4e00",OtherLetter),("4e03",OtherLetter),("4e07",OtherLetter),("4e09",OtherLetter),("4e5d",OtherLetter),("4e8c",OtherLetter),("4e94",OtherLetter),("4e96",OtherLetter),("4ebf",OtherLetter),("4ec0",OtherLetter),("4edf",OtherLetter),("4ee8",OtherLetter),("4f0d",OtherLetter),("4f70",OtherLetter),("5104",OtherLetter),("5146",OtherLetter),("5169",OtherLetter),("516b",OtherLetter),("516d",OtherLetter),("5341",OtherLetter),("5343",OtherLetter),("5344",OtherLetter),("5345",OtherLetter),("534c",OtherLetter),("53c1",OtherLetter),("53c2",OtherLetter),("53c3",OtherLetter),("53c4",OtherLetter),("56db",OtherLetter),("58f1",OtherLetter),("58f9",OtherLetter),("5e7a",OtherLetter),("5efe",OtherLetter),("5eff",OtherLetter),("5f0c",OtherLetter),("5f0d",OtherLetter),("5f0e",OtherLetter),("5f10",OtherLetter),("62fe",OtherLetter),("634c",OtherLetter),("67d2",OtherLetter),("6f06",OtherLetter),("7396",OtherLetter),("767e",OtherLetter),("8086",OtherLetter),("842c",OtherLetter),("8cae",OtherLetter),("8cb3",OtherLetter),("8d30",OtherLetter),("9621",OtherLetter),("9646",OtherLetter),("964c",OtherLetter),("9678",OtherLetter),("96f6",OtherLetter),("f96b",OtherLetter),("f973",OtherLetter),("f978",OtherLetter),("f9b2",OtherLetter),("f9d1",OtherLetter),("f9d3",OtherLetter),("f9fd",OtherLetter),("20001",OtherLetter),("20064",OtherLetter),("200e2",OtherLetter),("20121",OtherLetter),("2092a",OtherLetter),("20983",OtherLetter),("2098c",OtherLetter),("2099c",OtherLetter),("20aea",OtherLetter),("20afd",OtherLetter),("20b19",OtherLetter),("22390",OtherLetter),("22998",OtherLetter),("23b1b",OtherLetter),("2626d",OtherLetter),("2f890",OtherLetter)]d","4f70","5104","5146","5169","516b","516d","5341","5343","5344","5345","534c","53c1","53c2","53c3","53c4","56db","58f1","58f9","5e7a","5efe","5eff","5f0c","5f0d","5f0e","5f10","62fe","634c","67d2","6f06","7396","767e","8086","842c","8cae","8cb3","8d30","9621","9646","964c","9678","96f6","f96b","f973","f978","f9b2","f9d1","f9d3","f9fd","20001","20064","200e2","20121","2092a","20983","2098c","2099c","20aea","20afd","20b19","22390","22998","23b1b","2626d","2f890"]
```

File: lib/Unicode/Char/Numeric.hs (hunk 5 above)
Comment (wismill) [reply lib/Unicode/Char/Numeric.hs]: Done

File: lib/Unicode/Char/Numeric.hs (hunk 3 above)
Comment (wismill) [reply lib/Unicode/Char/Numeric.hs]: OK. I think I chose it for the resulting type.

File: lib/Unicode/Char/Numeric.hs (hunk 2 above)
Comment (wismill) [reply lib/Unicode/Char/Numeric.hs]: Yes. I updated the `.editorconfig` file.

File: lib/Unicode/Char/Numeric.hs [hunk 6]
```diff
@@ -21,50 +25,70 @@ module Unicode.Char.Numeric
...
+-- __Note:__ a character may have a numeric value but return 'False' with
+-- the predicate 'Unicode.Char.Numeric.Compat.isNumber', because
+-- 'Unicode.Char.Numeric.Compat.isNumber' only tests
+-- 'Unicode.Char.General.GeneralCategory': some CJK characters are
+-- 'Unicode.Char.General.OtherLetter' and do have a numeric value.
+--
+-- prop> hasNumericValue c == isJust (numericValue c)
+{-# INLINE hasNumericValue #-}
+hasNumericValue :: Char -> Bool
+hasNumericValue = isJust . V.numericValue
```
Comment (harendra-kumar) [inline lib/Unicode/Char/Numeric.hs:68]: We can call it `isNumeric`.

**Comment by wismill** [comment]
> Did you measure the perf of `isNumber` vs `hasNumericValue`? Maybe we can use a benchmark for that. The difference is looking up generalCategory vs looking up numeric value, I am not sure which one is faster and if there is a significant difference.

Yes: hasNumericValue is little faster. See the benchmark.

**Comment by wismill** [comment]
@harendra-kumar kind reminder for a review

File: lib/Unicode/Char/Numeric.hs (hunk 6 above)
Comment (harendra-kumar) [reply lib/Unicode/Char/Numeric.hs:68]: @wismill what is your opinion on renaming this to `isNumeric`? That's a shorter name and in line with other predicates.

**Review by harendra-kumar — APPROVED** [review]

File: lib/Unicode/Char/Numeric.hs (hunk 6 above)
Comment (wismill) [reply lib/Unicode/Char/Numeric.hs]: I agree. Done.

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**